agent/agent_endpoint.go

```diff
@@ -1487,3 +1490,90 @@ func (s *HTTPServer) AgentHost(resp http.ResponseWriter, req *http.Request) (int
 
 	return debug.CollectHostInfo(), nil
 }
+
+// AgentBootstrapGossipKey
+//
+// PUT /v1/agent/bootstrap-gossip-key
+//
+// Allows delivering the gossip encryption key via API to avoid it being present
+// on disk if -encrypt-key-from-api and -disable-keyring-file are used.
+func (s *HTTPServer) AgentBootstrapGossipKey(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	// Fetch the ACL token, if any, and enforce agent policy. We require
```

Comment:
If this is the very first time, will you even be able to dial the server nodes, or would this work with the agent master token instead?

Reply:
Hmmm. That's a good point. I didn't actually test end to end with ACL enabled and the "unit" tests for this run a `TestAgent` which is also a server.

I was in two minds about whether we needed this anyway but I think for clients it actually might be impossible in general to resolve the token until Gossip is up since the server pool won't be seeded yet.

I'll test that out and fix this.

Reply:
Well spotted. With default-allow ACL on a client agent this causes it to always 403.

```
    2018/12/03 13:01:59 [WARN] manager: No servers available
    2018/12/03 13:01:59 [ERR] http: Request PUT /v1/agent/bootstrap-gossip-key, error: Permission denied from=127.0.0.1:58208
    2018/12/03 13:01:59 [DEBUG] http: Request PUT /v1/agent/bootstrap-gossip-key (238.454µs) from=127.0.0.1:58208
```

Even if you deliver the agent token locally with the API and use that agent token in this call, this resolve path still errors on not being able to contact servers.

I'll remove the ACL check for now - as I said it shouldn't be necessary for security and there is a fundamental limit here - we have to trust the thing that is delivering keys since we can't have any shared secret with which to authenticate that thing until we are connected anyway. The fact it needs to know the correct gossip key is _implicitly_ the authorization needed to deliver that key I suppose.